test/chainbase_test.cpp

```diff
@@ -0,0 +1,506 @@
+                  upper_bound_inclusive,
+                  num_of_accounts_and_values,
+                  num_of_swaps,
+                  max_key_length,
+                  max_key_value,
+                  max_value_length,
+                  max_value_value}
+   {
+      _database.add_index<account_index>();
+   }
+
+   ~database_test()
+   {
+      boost::filesystem::remove_all(_database_dir);
+   }
+
+   inline void start_test() {
+      _initial_database_state();
+      _execution_loop();
+      loggerman.flush_all();
+   }
+
+private:
+   const boost::filesystem::path _database_dir{boost::filesystem::current_path() /= std::string{"/chainbase"}};
+   chainbase::database _database{_database_dir, chainbase::database::read_write, static_cast<uint64_t>((uint64_t)4096*(uint64_t)100000000)};
+   generated_data _gen_data;
+   system_metrics _system_metrics;
+
+   inline void _initial_database_state() {
+      size_t transactions_per_second{};
+
+      time_t old_time{initial_time};
+      time_t new_time{initial_time};
+      
+      std::cout << "Filling initial database state...\n" << std::flush;
+      loggerman.print_progress(1,0);
+      for (size_t i{}; i < _gen_data.num_of_accounts_and_values()/10; ++i) {
+         new_time = time(NULL);
+         if (new_time != old_time) {
+            loggerman.log_tps      ({(new_time - initial_time), transactions_per_second/(new_time - initial_time)});
+            loggerman.log_ram_usage({(new_time - initial_time), _system_metrics.total_ram_currently_used()});
+            loggerman.log_cpu_load ({(new_time - initial_time), _system_metrics.calculate_cpu_load()});
+            loggerman.print_progress(i, _gen_data.num_of_accounts_and_values()/10);
+            old_time = new_time;
+         }
+
+         _database.start_undo_session(true);
+         
+         auto session{_database.start_undo_session(true)};
+         _database.start_undo_session(true);
+         // Create 10 new accounts per undo session.
+         // AKA; create 10 new accounts per block.
+         for (size_t j{}; j < 10; ++j){
+            _database.create<account>([&](account& acc) {
+               acc._account_key   = _gen_data.accounts()[i*10+j];
+               acc._account_value = _gen_data.values  ()[i*10+j];
+            });
+         }
+         session.push();
+      }
+
+      loggerman.print_progress(1,1);
+      _database.start_undo_session(true).push();
+      std::cout << "done.\n" << std::flush;
+   }
+
+   inline void _execution_loop() {
+      size_t transactions_per_second{};
+
+      time_t old_time{initial_time};
+      time_t new_time{initial_time};
+
+      std::cout << "Benchmarking...\n" << std::flush;
+      loggerman.print_progress(1,0);
+      for (size_t i{}; i < _gen_data.num_of_swaps(); ++i) {
+         new_time = time(NULL);
+         if (new_time != old_time) {
+            loggerman.log_tps      ({(new_time - initial_time), transactions_per_second/(new_time - initial_time)});
```

Comment:
Did you mean to use `initial_time` here or `old_time` when calculating the TPS? The current system is determining TPS by averaging over an ever expanding time window. If you want to calculate the TPS within certain narrow time windows, you could use `transactions_per_second/(new_time - old_time)` instead but then you would also need to reset `transactions_per_second` each time you log the TPS. And if you want TPS averaged over a rolling time window, you will need to do more work.

And once again same comment as above applies here about having better control over when this data is logged as time advances.

Reply:
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L499-L503
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L631-L651
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L678-L688
The user now has the option to specify how to measure the test.